File: api/app/models/water_meter.rb
class WaterMeter < ApplicationRecord
  include SoftDeletable

  belongs_to :invoice, -> { unscope(where: :deleted_at) }

  validates :measured_at, :initial_reading, :final_reading, presence: true
  validates :initial_reading, :final_reading, numericality: { greater_than_or_equal_to: 0 }
  validate :final_reading_greater_than_initial

  FREE_TIER_LIMIT = 10000

  def consumption
    [ (final_reading - initial_reading) - FREE_TIER_LIMIT, 0 ].max
  end

  def real_consumption
    final_reading - initial_reading
  end

  alias_method :total_consumption, :real_consumption

  def water_consumed_value
    (consumption / 4.0) * 0.01
  end

  private

  def final_reading_greater_than_initial
    return if final_reading.blank? || initial_reading.blank?

    if final_reading < initial_reading
      errors.add(:final_reading, "must be greater than or equal to initial_reading")
    end
  end
end

File: app/lib/features/invoices/domain/water_meter.dart
class WaterMeter {
  final String? id;
  final String invoiceId;
  final double initialReading;
  final double finalReading;
  final DateTime measuredAt;

  const WaterMeter({
    this.id,
    required this.invoiceId,
    required this.initialReading,
    required this.finalReading,
    required this.measuredAt,
  });

  double get consumption => (finalReading - initialReading) - 10000 > 0
      ? (finalReading - initialReading) - 10000
      : 0;

  double get realConsumption => finalReading - initialReading;

  factory WaterMeter.fromJson(Map<String, dynamic> json) => WaterMeter(
        id: json['id']?.toString(),
        invoiceId: json['invoice_id'].toString(),
        initialReading: double.parse(json['initial_reading'].toString()),
        finalReading: double.parse(json['final_reading'].toString()),
        measuredAt: DateTime.parse(json['measured_at'] as String),
      );
}
